File: aeron-driver/src/test/cpp/status/SystemCountersTest.cpp
/*
 * Copyright 2014 - 2016 Real Logic Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <gtest/gtest.h>
#include <concurrent/AtomicCounter.h>
#include <concurrent/AtomicBuffer.h>
#include <concurrent/Atomic64.h>
#include <status/SystemCounters.h>

using namespace testing;
using namespace aeron::concurrent;
using namespace aeron::driver::status;

static const std::int32_t VALUE_BUFFER_LENGTH = 16 * 1024;
static const std::int32_t META_BUFFER_LENGTH = 2 * VALUE_BUFFER_LENGTH;

typedef std::array<std::uint8_t, VALUE_BUFFER_LENGTH> value_buffer_t;
typedef std::array<std::uint8_t, META_BUFFER_LENGTH> meta_buffer_t;

class SystemCountersTest : public Test
{
public:
    SystemCountersTest() :
        m_meta_buffer(&m_meta[0], m_meta.size()),
        m_value_buffer(&m_value[0], m_value.size()),
        m_countersManager(m_meta_buffer, m_value_buffer),
        m_systemCounters(m_countersManager)
    {}

    AERON_DECL_ALIGNED(meta_buffer_t m_meta, 16);
    AERON_DECL_ALIGNED(value_buffer_t m_value, 16);
    AtomicBuffer m_meta_buffer;
    AtomicBuffer m_value_buffer;
    CountersManager m_countersManager;
    SystemCounters m_systemCounters;
};

static void getAndSetCounter(SystemCounters& counters, SystemCounterDescriptor descriptor)
{
    std::int64_t value = 12345;

    AtomicCounter* c1 = counters.get(descriptor);
    c1->setOrdered(value);

    AtomicCounter* c2 = counters.get(descriptor);

    std::int64_t v2 = c2->get();
    ASSERT_EQ(value, v2);
}

TEST_F(SystemCountersTest, readAndWriteSystemCounter)
{
    for (auto descriptor : SystemCounterDescriptor::VALUES)
    {
        getAndSetCounter(m_systemCounters, descriptor);
    }
}

File: aeron-system-tests/src/test/java/io/aeron/driver/SystemCountersTest.java
/*
 * Copyright 2014-2025 Real Logic Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.aeron.driver;

import io.aeron.Aeron;
import io.aeron.driver.status.SystemCounterDescriptor;
import io.aeron.test.SystemTestWatcher;
import io.aeron.test.driver.TestMediaDriver;
import org.agrona.CloseHelper;
import org.agrona.collections.Int2ObjectHashMap;
import org.agrona.concurrent.status.CountersReader;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.RegisterExtension;

import static org.hamcrest.CoreMatchers.startsWith;
import static org.hamcrest.MatcherAssert.assertThat;

public class SystemCountersTest
{
    @RegisterExtension
    final SystemTestWatcher systemTestWatcher = new SystemTestWatcher();

    private TestMediaDriver driver;
    private Aeron aeron;

    @BeforeEach
    void before()
    {
        final MediaDriver.Context context = new MediaDriver.Context()
            .dirDeleteOnStart(true)
            .dirDeleteOnShutdown(true)
            .threadingMode(ThreadingMode.SHARED);
        driver = TestMediaDriver.launch(context, systemTestWatcher);
        systemTestWatcher.dataCollector().add(driver.context().aeronDirectory());

        aeron = Aeron.connect(new Aeron.Context().aeronDirectoryName(context.aeronDirectoryName()));
    }

    @AfterEach
    void after()
    {
        CloseHelper.closeAll(aeron, driver);
    }

    @Test
    void shouldCreatePublicationUsingSparseFiles()
    {
        final CountersReader countersReader = aeron.countersReader();
        final Int2ObjectHashMap<String> idToLabel = new Int2ObjectHashMap<>();
        countersReader.forEach((counterId, typeId, keyBuffer, label) ->
        {
            if (SystemCounterDescriptor.SYSTEM_COUNTER_TYPE_ID == typeId)
            {
                idToLabel.put(keyBuffer.getInt(0), label);
            }
        });

        for (final SystemCounterDescriptor counter : SystemCounterDescriptor.values())
        {
            final String counterLabel = idToLabel.get(counter.id());
            assertThat(counterLabel, startsWith(counter.label()));
        }
    }
}
